File: python_code/stock_profile.py
import httpx

url = 'https://stockanalysis.com/stocks/smci/company/__data.json'
response = httpx.get(url)
raw_data = response.json()

nodes = raw_data['nodes']
data = nodes[2]['data']
data_map = data[0]

description_index = data_map['description']
stock_profile_index = data_map['profile']
stock_profile_data = data[stock_profile_index]
stock_industry_data_index = stock_profile_data['industry']
stock_sector_data_index = stock_profile_data['sector']

stock_name_index = stock_profile_data['name']
stock_country_index = stock_profile_data['country']
stock_founded_index = stock_profile_data['founded']
stock_ipo_date_index = stock_profile_data['ipoDate']
stock_ceo_index = stock_profile_data['ceo']
stock_industry_index = data[stock_industry_data_index]['value']
stock_sector_index = data[stock_sector_data_index]['value']

stock_profile_doc = {
    'name': data[stock_name_index],
    'description': data[description_index],
    'country': data[stock_country_index],
    'founded': data[stock_founded_index],
    'ipoDate': data[stock_ipo_date_index],
    'industry': data[stock_industry_index],
    'sector': data[stock_sector_index],
    'ceo': data[stock_ceo_index],
}

print(stock_profile_doc)

File: marketDataScraper/stock_profile.go
package marketDataScraper

import (
	"encoding/json"
	"fmt"
	"investbot/domain"
	"io"
	"net/http"
)

func GetStockProfile(symbol string) (domain.StockProfile, error) {
	url := fmt.Sprintf("https://stockanalysis.com/stocks/%s/company/__data.json", symbol)

	resp, err := http.Get(url)
	if err != nil {
		return domain.StockProfile{}, err
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return domain.StockProfile{}, err
	}

	var rawData map[string]interface{}
	if err := json.Unmarshal(body, &rawData); err != nil {
		return domain.StockProfile{}, err
	}

	// Extract "nodes" from rawData
	nodes, ok := rawData["nodes"].([]interface{})
	if !ok || len(nodes) < 3 {
		return domain.StockProfile{}, fmt.Errorf("unexpected structure in 'nodes'")
	}

	// Access the second element in "nodes" which contains the data we are interested in
	nodeData, ok := nodes[2].(map[string]interface{})
	if !ok {
		return domain.StockProfile{}, fmt.Errorf("unexpected structure in 'nodes[2]'")
	}

	data, ok := nodeData["data"].([]interface{})
	if !ok {
		return domain.StockProfile{}, fmt.Errorf("unexpected structure in 'data'")
	}

	dataMap, ok := data[0].(map[string]interface{})
	if !ok {
		return domain.StockProfile{}, fmt.Errorf("unexpected structure in 'data[0]'")
	}

	descriptionIndex, ok := dataMap["description"].(float64)
	if !ok {
		return domain.StockProfile{}, fmt.Errorf("unexpected structure for 'description'")
	}

	profileIndex, ok := dataMap["profile"].(float64)
	if !ok {
		return domain.StockProfile{}, fmt.Errorf("unexpected structure for 'profile'")
	}

	stockProfileData := data[int(profileIndex)].(map[string]interface{})

	industryDataIndex, ok := stockProfileData["industry"].(float64)
	if !ok {
		return domain.StockProfile{}, fmt.Errorf("unexpected structure for 'industry'")
	}

	industryData := data[int(industryDataIndex)].(map[string]interface{})
	industryNameIndex := industryData["value"].(float64)

	sectorDataIndex, ok := stockProfileData["sector"].(float64)
	if !ok {
		return domain.StockProfile{}, fmt.Errorf("unexpected structure for 'sector'")
	}
	sectorData := data[int(sectorDataIndex)].(map[string]interface{})
	sectorNameIndex := sectorData["value"].(float64)

	stockNameINdex := stockProfileData["name"].(float64)
	stockCountryIndex := stockProfileData["country"].(float64)
	stockFoundedIndex := stockProfileData["founded"].(float64)
	stockIpoDateIndex := stockProfileData["ipoDate"].(float64)
	stockCeoIndex := stockProfileData["ceo"].(float64)

	return domain.StockProfile{
		Name:        data[int(stockNameINdex)].(string),
		Description: data[int(descriptionIndex)].(string),
		Country:     data[int(stockCountryIndex)].(string),
		Founded:     int(data[int(stockFoundedIndex)].(float64)),
		IpoDate:     data[int(stockIpoDateIndex)].(string),
		Industry:    data[int(industryNameIndex)].(string),
		Sector:      data[int(sectorNameIndex)].(string),
		Ceo:         data[int(stockCeoIndex)].(string),
	}, nil
}
